### ISSUE TITLE
what's the motivation behind forkFinally?

### ISSUE THREAD

**Issue by qrilka** [nikita-volkov/slave-thread#5]
From the docs it's not quite clear what's the difference between ``fork (action `finally` finalizer)`` and `forkFinally finalizer action` - any hints on that?

**Comment by nikita-volkov** [comment]
Hierarchical releasing of resources. The deeper-nested forked threads are guaranteed to execute their finalizers first.

**Comment by qrilka** [comment]
So the main point is about the order of "finalizers"? Could you name some example where fork+finally will give you not correct order?
It looks to me like a good thing to add into haddocks but I'd like to understand the point to document it correctly (or you could do that yourself of course if you have will and time)

**Comment by nikita-volkov** [comment]
I've actually never tested, whether the same could be achieved with fork+finally. Also I implemented this library too long ago to remember the roots of my design decisions :)

**Comment by qrilka** [comment]
Then feel free to close the ticket @nikita-volkov

**Comment by mitchellwrosen** [comment]
@nikita-volkov Any interest in revisiting this design? I've looked over the code and I noticed it's possible for a thread to die without running its finalizer, and without deregistering itself.

```haskell
import Control.Concurrent (threadDelay)
import Control.Exception
import SlaveThread
import System.IO

main :: IO ()
main = do
  forkFinally finalizer1 computation1
  threadDelay 1000000

finalizer1 = do
  putStrLn "[thread 1] finalizer"

computation1 = do
  putStrLn "[thread 1] running"
  putStrLn "[thread 1] spawning thread 2"
  forkFinally finalizer2 computation2

finalizer2 = do
  putStrLn "[thread 2] finalizer"
  throwIO (userError "finalizer2")

computation2 = do
  putStrLn "[thread 2] running"
```

In the code above, `finalizer1` will not run.

This is because:

- Thread 1 runs, and when `computation1` finishes, we kill all child threads.
- So `finalizer2` runs, which throws the `userError` back to `thread1`.
- Thread 1 waits for its children to die, which is an interruptible action. So the `userError` is delivered to thread 1 before it gets a chance to run its finalizer.

**Comment by mitchellwrosen** [comment]
And note that if thread 1 instead spawns thread 2 as

```haskell
fork (finally finalizer2 computation2)
```

this issue does not arise.

**Comment by mitchellwrosen** [comment]
@qrilka

```
fork (action `finally` cleanup)
```

does not guarantee that `cleanup` runs after all childrens' finalizers. In fact, the opposite - it would always run `cleanup` _before_ any child finalizers.

**Comment by nikita-volkov** [comment]
@mitchellwrosen Great catch! Pun unintended :) It's a bug. I'll fix that.

**Comment by nikita-volkov** [comment]
Fixed in 1.0.2.3

**Comment by nikita-volkov** [comment]
@qrilka Concerning your original question, see my answer on Reddit: https://www.reddit.com/r/haskell/comments/2jo3ty/announce_slave_thread_a_solution_to_ghost_threads/cldodfv/
